### ISSUE TITLE
Update home-assistant network Thread credentials not possible

### ISSUE THREAD

**Issue by agners** [home-assistant/android#4146]
**Home Assistant Android app version(s):**
2024.1.2-full

**Android version(s):**
14

**Device model(s):**
Pixel Pro 8

**Home Assistant version:**
2024.1.3

**Last working Home Assistant release (if known):**

**Description of problem, include YAML if issue is related to notifications:**
When creating a new `home-assistant` Thread network in Home Assistant (e.g. via "Reset border router" feature in the Home Assistant Thread integration configuration page) updating the Thread credentials seems not to work. 

To reproduce:

1. "Clear all data" of "Google Play services" method to make sure to start from a clear slate
2. Have a `home-assistant` network as preferred set in Core (say `home-assistant` network 1)
3. Select "Sync Thread credentials" -> "✅ Added network from Home Assistant to this device"
4. Select "Sync Thread credentials" -> "✅ Home Assistant and this device use the same network" :+1: 
5. Create a `home-assistant` Thread network e.g. by using "Reset border router"
6. Select "Make preferred network" to mark this new `home-assistant` network as preferred
7. Select "Sync Thread credentials" -> "✅ Updated network from Home Assistant to this device"
8. Select "Sync Thread credentials" -> "✅ Updated network from Home Assistant to this device"

Up to 7 everything is as expected. However, when trying to commission a device after step 7, it won't work because the credentials didn't really got updated. Step 8 also proves that. It seems the app is not able to update the dataset.

<details>

**Screenshot and log of step 3:**

![image](https://github.com/home-assistant/android/assets/34061/31e4fcad-c02f-4747-8206-dcd79be0a47d)

```
01-20 14:36:30.259  7110  7110 W WindowOnBackDispatcher: sendCancelIfRunning: isInProgress=falsecallback=com.google.android.material.motion.MaterialBackOrchestrator$Api34BackCallbackDelegate$1@288a0ce
01-20 14:36:30.268  7110  7110 W InputEventReceiver: Attempted to finish an input event but the input event receiver has already been disposed.
01-20 14:36:30.282  7110  7110 D ServerConnectionInfo: localUrl is: false, usesInternalSsid is: false, usesWifi is: true
01-20 14:36:30.282  7110  7110 D WebSocketRepository: Sending message 49: {type=get_config, id=49}
01-20 14:36:30.282  7110  7110 D WebSocketRepository: Message number 49 sent
01-20 14:36:30.343  7110 13794 D WebSocketRepository: Websocket: onMessage (text)
01-20 14:36:30.344  7110 13794 D WebSocketRepository: Message number 49 received
01-20 14:36:30.365  7110  7110 D WebSocketRepository: Sending message 50: {type=thread/list_datasets, id=50}
01-20 14:36:30.365  7110  7110 D WebSocketRepository: Message number 50 sent
01-20 14:36:30.372  7110 13794 D WebSocketRepository: Websocket: onMessage (text)
01-20 14:36:30.373  7110 13794 D WebSocketRepository: Message number 50 received
01-20 14:36:40.384  7110  7110 D WebSocketRepository: Sending message 51: {type=thread/get_dataset_tlv, dataset_id=01H7SF0EZ6CZR08N4P2G77R6ZG, id=51}
01-20 14:36:40.384  7110  7110 D WebSocketRepository: Message number 51 sent
01-20 14:36:40.396  7110 13794 D WebSocketRepository: Websocket: onMessage (text)
01-20 14:36:40.397  7110 13794 D WebSocketRepository: Message number 51 received
01-20 14:36:40.472  7110  7110 D ThreadManagerImpl: Thread import to device completed
01-20 14:36:49.002  7110  7110 W WindowOnBackDispatcher: sendCancelIfRunning: isInProgress=falsecallback=android.app.Dialog$$ExternalSyntheticLambda2@e6e29be
01-20 14:36:49.018  7110  7110 W InputEventReceiver: Attempted to finish an input event but the input event receiver has already been disposed.
01-20 14:36:49.031  7110  7110 D IntegrationRepository: isAppLocked(): false. (LockEnabled: false, appActive: true, expireMillis: 0, currentMillis: 1705757809031)
01-20 14:36:49.031  7110  7110 D IntegrationRepository: setAppActive(): true
01-20 14:36:49.891  7110  7277 D LogcatReader: Read logcat for pid 7110
```

**Screenshot and log of step 8:**

![image](https://github.com/home-assistant/android/assets/34061/eeedb5c4-bd87-4471-a71f-9380070b8043)

```
01-20 14:39:22.718  7110  7110 W WindowOnBackDispatcher: sendCancelIfRunning: isInProgress=falsecallback=com.google.android.material.motion.MaterialBackOrchestrator$Api34BackCallbackDelegate$1@53b36cf
01-20 14:39:22.730  7110  7110 W InputEventReceiver: Attempted to finish an input event but the input event receiver has already been disposed.
01-20 14:39:22.732  7110  7110 D ServerConnectionInfo: localUrl is: false, usesInternalSsid is: false, usesWifi is: true
01-20 14:39:22.733  7110  7110 D WebSocketRepository: Sending message 59: {type=get_config, id=59}
01-20 14:39:22.733  7110  7110 D WebSocketRepository: Message number 59 sent
01-20 14:39:22.742  7110 13794 D WebSocketRepository: Websocket: onMessage (text)
01-20 14:39:22.742  7110 13794 D WebSocketRepository: Message number 59 received
01-20 14:39:22.757  7110  7110 D WebSocketRepository: Sending message 60: {type=thread/list_datasets, id=60}
01-20 14:39:22.757  7110  7110 D WebSocketRepository: Message number 60 sent
01-20 14:39:22.768  7110 13794 D WebSocketRepository: Websocket: onMessage (text)
01-20 14:39:22.768  7110 13794 D WebSocketRepository: Message number 60 received
01-20 14:39:32.788  7110  7110 D WebSocketRepository: Sending message 61: {type=thread/get_dataset_tlv, dataset_id=01HMKD0FR3Q9J27680PM3293DQ, id=61}
01-20 14:39:32.789  7110  7110 D WebSocketRepository: Message number 61 sent
01-20 14:39:32.797  7110 13794 D WebSocketRepository: Websocket: onMessage (text)
01-20 14:39:32.797  7110 13794 D WebSocketRepository: Message number 61 received
01-20 14:39:42.843  7110  7110 D ThreadManagerImpl: Thread: device doesn't prefer core preferred dataset
01-20 14:40:02.974  7110  7110 D ThreadManagerImpl: Thread device prefers app added dataset: home-assistant (PAN 19939, EXTPAN /��xɎ)
01-20 14:40:02.974  7110  7110 D ThreadManagerImpl: Thread: device prefers dataset from app
01-20 14:40:02.975  7110  7110 D WebSocketRepository: Sending message 62: {type=thread/get_dataset_tlv, dataset_id=01HMKD0FR3Q9J27680PM3293DQ, id=62}
01-20 14:40:02.975  7110  7110 D WebSocketRepository: Message number 62 sent
01-20 14:40:02.997  7110 13794 D WebSocketRepository: Websocket: onMessage (text)
01-20 14:40:02.998  7110 13794 D WebSocketRepository: Message number 62 received
01-20 14:40:03.064  7110  7110 D ThreadManagerImpl: Thread update device completed: deleted 1 datasets, updated 1
01-20 14:40:10.140  7110  7110 W WindowOnBackDispatcher: sendCancelIfRunning: isInProgress=falsecallback=android.app.Dialog$$ExternalSyntheticLambda2@d20115
01-20 14:40:10.146  7110 11187 D OpenGLRenderer: endAllActiveAnimators on 0xb4000077e690e3a0 (RippleDrawable) with handle 0xb400007846579d40
01-20 14:40:10.170  7110  7110 D IntegrationRepository: isAppLocked(): false. (LockEnabled: false, appActive: true, expireMillis: 1705757942197, currentMillis: 1705758010170)
01-20 14:40:10.170  7110  7110 D IntegrationRepository: setAppActive(): true
01-20 14:40:14.569  7110  7110 D IntegrationRepository: isAppLocked(): false. (LockEnabled: false, appActive: true, expireMillis: 1705757942197, currentMillis: 1705758014569)
01-20 14:40:14.569  7110  7110 D IntegrationRepository: setAppActive(): true
01-20 14:40:14.811  7110  7174 D LogcatReader: Read logcat for pid 7110
```

**Additional information:**

**Comment by agners** [comment]
The current work around is to delete all Thread credentials on the device. The Thread credentials are stored as part of the Google Play services. They ca be removed using the "Clear all data" button in the Google Play service storage menu.

:warning:  Deleting Google Play services data does reset some other device settings, namely the Google Wallet cards. :warning: 

![Screenshot_20231213-105945](https://github.com/home-assistant/android/assets/34061/bf098177-4477-4111-a3c4-60dba4fcd447)

**Comment by jpelgrom** [comment]
Potential library bug: https://issuetracker.google.com/issues/286158210

The example app also doesn't do anything for clearing the credential which is surprising: https://github.com/google-home/sample-apps-for-matter-android/blob/244ccdd78b7e869fe5260adbe0d051e8c9b5ee29/3p-ecosystem/src/main/java/com/google/homesampleapp/screens/thread/ThreadViewModel.kt#L133-L135

I have not yet tried to replicate this because of a Nest Hub on my network which I need to reset + keep offline while doing this to prevent it from becoming the preferred credential. Maybe/hopefully logcat on device will provide a hint what is happening inside GPS, even if it is redacted.

**Comment by WindFreaker** [comment]
Are there any updates on this issue? No matter what I do I can't get my Android phone to prefer the HA thread border router, which causes a conflict and blocks adding new devices.

**Comment by jimb0b** [comment]
Hi - I found this thread as I was also having issues with syncing Thread credentials to Android from the Thread integration page (pressing the Import Credentials button) - no matter what I did (including wiping my google play services data) it would not sync. I even tried deleting the OpenThread and Thread integrations from HA and setting everything up from scratch, with no avail.

Then I found another thread (https://community.home-assistant.io/t/matter-over-thread-commissioning-issue-with-gl-inet-s200/695901/2) which talks about a different method of syncing credentials, specifically via the Android app in "Settings -> Companion App -> Troubleshooting -> Sync thread credentials" route. This method of syncing worked for me! It still requires trashing google play services data unfortunately, so it's more of a workaround - but wanted to post it here in case this is the step others are struggling with.

**Comment by j-m** [comment]
The workaround also worked for me, however it's worth highlighting a few extra things.

1. Deleting your google play service data also removes things like Find My Device. Pretend you have a new phone and double check everything is as it was!
2. Only your credit/debit cards in your Google Wallet will need reverifying via their CVCs (or completely re-added if not saved to your google account), whereas things like shop memberships etc will be kept. 
3. If you managed to connect your device to thread but you're still having trouble connecting the device to matter, try disconnecting from any VPNs. I have tailscale on my HA and had it set on my android as an Always On vpn and to block any non-vpn connections, and for whatever reason this prevented matter from successfully generating new credentials.
4. Somewhat annoyingly, my fiancée's iPhone had no problems whatsoever changing preferred networks. So, if you're struggling, try the dark side!

My scenario was for Nanoleaf A19s/NL45s & NL65s, but I figure this could apply more broadly. Hope this helps someone!

**Comment by linux4life798** [comment]
I ran into a very similar issue, where my Pixel 8 Pro kept trying to transfer Thread credentials from an old deleted SmartThings Hub network instead of my current Nest Thread network. What's strange is that I had just successfully commissioned a Matter device through Google Home on the correct PAN and then shared it with Home Assistant, but the phone still wouldn't transfer the correct credentials.

I found a workaround by using a different phone (Pixel 9) that had never set up any Matter devices before. Though I initially got some "Thread not available" errors, after playing with the local device Matter settings for a while (requested details for existing matter devices on the Nest PAN), I was finally able to transfer the correct Nest network credentials.

**Comment by colonelpanic8** [comment]
The trick of deleting google play services data definitely worked for me, but I'm now encountering a new issue. The message simply says "Can't reach device" with the subheading:  "Make sure your phone is connected to Wi-Fi"

**Comment by brettmaustin** [comment]
> I ran into a very similar issue, where my Pixel 8 Pro kept trying to transfer Thread credentials from an old deleted SmartThings Hub network instead of my current Nest Thread network. What's strange is that I had just successfully commissioned a Matter device through Google Home on the correct PAN and then shared it with Home Assistant, but the phone still wouldn't transfer the correct credentials.
> 
> I found a workaround by using a different phone (Pixel 9) that had never set up any Matter devices before. Though I initially got some "Thread not available" errors, after **playing with the local device Matter settings for a while (requested details for existing matter devices on the Nest PAN), I was finally able to transfer the correct Nest network credentials**.

Can you please elaborate on what you meant by the bolded text above? I cannot get a new Google Streamer to accept my Thread credentials, even after deleting Google Play Services data and then forcing a sync of my Thread credentials in the Companion App.

**Comment by linux4life798** [comment]
> > I ran into a very similar issue, where my Pixel 8 Pro kept trying to transfer Thread credentials from an old deleted SmartThings Hub network instead of my current Nest Thread network. What's strange is that I had just successfully commissioned a Matter device through Google Home on the correct PAN and then shared it with Home Assistant, but the phone still wouldn't transfer the correct credentials.
> > I found a workaround by using a different phone (Pixel 9) that had never set up any Matter devices before. Though I initially got some "Thread not available" errors, after _**playing with the local device Matter settings for a while (requested details for existing matter devices on the Nest PAN), I was finally able to transfer the correct Nest network credentials**._
> 
> Can you please elaborate on what you meant by the bolded/italicized text above? I cannot get a new Google Streamer to accept my Thread credentials, even after deleting Google Play Services data and then forcing a sync of my Thread credentials in the Companion App.

Yep. To be clear, the goal was simply to transfer the already created Nest Thread PAN credentials to the Home Assistant thread services, so that I could get the home assistant on the same Thread network/PAN as Nest. This is because I have not found a way  to have the Nest routers join the Thread network/PAN that home assistant creates.

On the Pixel 9, which had never commissioned any Matter device, I simply navigated into the settings for a particular Matter device in the Google Home app and waited a bit. Looking at Google Home, this may have been simply viewing the "Linked Matter app & service" and/or the "Technical information" items. Note that I previously added this Matter+thread (probably some Eve energy device) directly to Google Home using my Pixel 8 and then shared the device with home assistant.

**Comment by linux4life798** [comment]
> The trick of deleting google play services data definitely worked for me, but I'm now encountering a new issue. The message simply says "Can't reach device" with the subheading: "Make sure your phone is connected to Wi-Fi"

From what I recall, **reinstalling/deleting data for google play services is risky**, due to other apps relying on it to be in a consistent state. When I cleared the storage for it, a while ago, all of my Google Wallet cards were deleted and other apps had to be reset too, due to lots of odd errors and malfunctions. The most problematic thing was the deletion of my Clipper card (public transit), since there is no way to recover it. Had to pay $5 to transfer fund to new card.

**Comment by itpeters** [comment]
I don't know if this is new, but at least on my Pixel phone running Android 16 (BP2A.250605.031.A2, along with Google Play services 25.26.31 (260400-776620933)), when I search for "thread" in the settings app, there's a result called "Thread networks" provided by Google Play services.

Once I open up Thread networks, it has what looks like a list of available networks, but it only seems to know about the Google-run network from my Nest Hub (screenshot attached). It also indicates this is the preferred network, which implies (to me at least!) that this might be where it is, or will become, possible to change the preferred network without clearing storage.

Unfortunately, it looks like there's no way to convince the Home Assistant companion app to provide the necessary credentials to have that network available. It looks like the API for that may exist (https://developers.home.google.com/thread), however.

Apologies if this isn't new information, and hopefully it leads somewhere! I'm very stubbornly refusing to wipe Google Play services and, for the time being, I'm able to add my Matter over Thread devices to my Home Assistant thread network using a spare Android tablet which has the correct preferred network, but obviously it would be great if there were a way around this issue.

EDIT: Oddly, and perhaps not encouragingly, the same Thread networks screen on the spare tablet only lists the Nest PAN and still doesn't list the Home Assistant thread network, even though it has been successfully adding devices to that PAN. The only difference is that it does not claim the Nest PAN to be "Preferred" on that device.

![Image](https://github.com/user-attachments/assets/a07a1283-3bb1-49ba-a408-d5fe063940ba)

**Comment by jpelgrom** [comment]
> I don't know if this is new, but (...) when I search for "thread" in the settings app, there's a result called "Thread networks" provided by Google Play services.

That panel is new, nice find! (And finally they added something like this) 

> Unfortunately, it looks like there's no way to convince the Home Assistant companion app to provide the necessary credentials to have that network available.

The Play Services Thread component used to ignore any additional networks added / there is no way in the API to make another network preferred, which is why the app currently won't try to add another network to the device if it already has one. But this interface suggests that might be changing so worth revisiting.

**Comment by miawgogo** [comment]
I was just troubleshooting something related to this, and i might raise it in the OTBR addon repo instead, but i have just noticed a disparity between how the OTBR addon responds to a `_meshcop._udp.local` request from android and how a certified product does (in this case its a nanoleaf shapes).

android sends out a request for `_meshcop._udp.local` when its scanning  for border routers on that thread settings dialog @itpeters shared, or when you sync keys from home assistant. One diffrence i have noticed is that home assistants border router is not responding with a txt record containing infomation about the thread network where as my nanoleaf router is

<details><summary>Response from Home assistant</summary>

<img width="1914" height="616" alt="Image" src="https://github.com/user-attachments/assets/c6b0e756-b84a-4bde-8dba-6dc99d829405" />

</details>

<details><summary>Response from Nanoleaf</summary>

<img width="1900" height="654" alt="Image" src="https://github.com/user-attachments/assets/4ab8e724-82ba-4296-976f-e7a8038fbb50" />

</details>

i wonder if google play services might not be adding the credentials as it expects to get some details from the TXT record that home assistant does not seem to be sending in response to its request(it will however send it in response to network tools 🤷)


edit: this behaviour seems pervasive throughout the matter related addons as there is a bug report about something simular happing with the matter commissioning https://github.com/home-assistant/addons/issues/4035

edit 2: for home assistant org members, i will send you the `.pcapng` files from wireshark if you need them uncensored(the only reason im limiting the offer) to compare to other border routers

**Comment by miawgogo** [comment]
Ah, the missing txt record is important for [adding credentials](https://developers.home.google.com/thread#add_credentials) with Googles Thread SDK

> Thread Border Routers must use mDNS to advertise Thread network information, including the Network Name, Extended Pan ID, and Border Agent ID. The corresponding txt values for these attributes are nn, xp, and id, respectively.

These do eventally get advertised, passively, but i think the fact they are missing during active requests might be related to the issues with adding credentials

**Comment by miawgogo** [comment]
this is about the extent i am able to test as i cant add this new border router anywhere atm. But setting up OTBR on a raspberry pi running the latest raspbian does let it show up in the "Thread Networks" settings and it includes the TXT record

<details><summary>wireshark</summary>
<img width="1876" height="860" alt="Image" src="https://github.com/user-attachments/assets/29fea758-9b54-4ce7-a6d6-f044c713c508" />
</details>

<details><summary>android thread settings</summary>

<img width="1440" height="3078" alt="Image" src="https://github.com/user-attachments/assets/af654d33-a80d-4471-a4df-b3bb40117e7a" />

</details>

there is a caveat with this test, its based on the latest main branch of the OTBR so its not exactly the same as the add-on, i did try mimicing the otbr addon, but i was having issues building the otbr that way.

Edit: Expanding on the above comment, without Google Play services being able to find a network, it just has a key without a corrisponding network, so it cant really manage it

Edit 2: Further investigation will be posted in the addons issues bellow

**Comment by miawgogo** [comment]
further updated on the MDNS stuff, this issue still persists, even with working `_meshcop._udp.local` announcements

**Comment by DomZZ** [comment]
Hello, we are in 2026 and everything (HA, Phone, coordinator) is uptodate but I'm still facing this issue with my pixel 7 pro and my Pixel 3, HA tread network never appears in phone settings ...

Will try tonight with wife iPhone ...

**Comment by albertgl89** [comment]
Same on Pixel 8...

**Comment by simowilso** [comment]
Just come across this issue. If I delete Play Store data, can I restore the Wallet information from backup? Or will that also restore the old Thread data?

**Comment by rogervn** [comment]
Same issue for me, but even clearing all google play services data didn't work. It did change the network, but it would stay forever in "Checking network connectivity".

I managed to add the device with an iPad, so it does seem it works (although it did take a few minutes and I thought it was stuck as well).

Did ANYONE get it to work reliably with Android?

**Comment by simowilso** [comment]
It worked for me, Google Pixel 8a

**Comment by Gunni** [comment]
Why does it ask me to connect to a 2.4 GHz Wi-Fi network when pairing Matter Over Thread devices?

**Comment by EmKayP** [comment]
In my scenario, Adguard home seems to be causing issues with mDNS. Disabling it allows me to see the Thread network and allows me to add the devices. I'll see what I can find about it in the next couple of days.

**Comment by zachd** [comment]
Note I had to unplug my NEST HUB device, and disable "Save credentials" in the Thread Networks page of the Android Settings app.

When removing all Google Play Services storage, immediately go to Home Assistant Settings -> Companion App -> Troubleshooting -> Sync Thread Credentials and do this twice for it to be setup correctly

**Comment by Gunni** [comment]
@zachd "Save credentials" just means "Send network details including secret key to Google"

**Comment by Downfor-u** [comment]
> The trick of deleting google play services data definitely worked for me, but I'm now encountering a new issue. The message simply says "Can't reach device" with the subheading: "Make sure your phone is connected to Wi-Fi"

Same for me. I could manage to advertise the newly provisioned thread network to my devices (after deleting google play services data) but now I have this message.
Did you find a solution ?

**Comment by eco9898** [comment]
> In my scenario, Adguard home seems to be causing issues with mDNS. Disabling it allows me to see the Thread network and allows me to add the devices. I'll see what I can find about it in the next couple of days.

@EmKayP did you find anything related to this. I also use adguard DNS and am experiencing issues connecting the device to the thread network.

**Comment by EmKayP** [comment]
> > In my scenario, Adguard home seems to be causing issues with mDNS. Disabling it allows me to see the Thread network and allows me to add the devices. I'll see what I can find about it in the next couple of days.
> 
> [@EmKayP](https://github.com/EmKayP) did you find anything related to this. I also use adguard DNS and am experiencing issues connecting the device to the thread network.

There are two GitHub issue discussions I've been participating on this, I think I updated what happened on the other thread. In my case, it wasn't AdGuard. My ISP router's DHCP options did not allow me to configure custom DNS server, so I "worked-around" the problem by using a light weight DHCP server that runs on a windows 11. Turns out, that doesn't support IPv6 (and disabling DHCP on the router also apparently turned of the IPv6 support on LAN).

I've since set up dnsmasq on an Ubuntu. That seems to have fixed all the issues that I had with the Thread network.

**Comment by purpledesign** [comment]
I also have this bug.

I got around it by setting up a second user on my phone. It appears that an initial troubleshoot > sync thread credentials works and then android locks that thread credential on the phone. Subsequent 'successful' credential syncs don't do anything.

**Comment by Furegato** [comment]
> I also have this bug.
> 
> I got around it by setting up a second user on my phone. It appears that an initial troubleshoot > sync thread credentials works and then android locks that thread credential on the phone. Subsequent 'successful' credential syncs don't do anything.

THIS!
Amazing idea! Worked greatly for me!
Thank you!
Finally a good workaround that don't need a second phone or a total reset of Google play which can mess everything around...
again... Thanks!
